tools/sgcloudspanner/emulator.go

```diff
@@ -0,0 +1,161 @@
+package sgcloudspanner
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"net"
+	"os"
+	"os/exec"
+	"strings"
+	"time"
+
+	"go.einride.tech/sage/sg"
+)
+
+// RunEmulator runs the Cloud Spanner emulator in Docker.
+func RunEmulator(ctx context.Context) (_ func(), err error) {
+	defer func() {
+		if err != nil {
+			err = fmt.Errorf("run Cloud Spanner emulator: %w", err)
+		}
+	}()
+	sg.Logger(ctx).Println("starting Cloud Spanner emulator...")
+	if emulatorHost, ok := os.LookupEnv("SPANNER_EMULATOR_HOST"); ok {
+		sg.Logger(ctx).Printf("a Cloud Spanner emulator is already running on %s", emulatorHost)
+		return func() {}, nil
+	}
+	if !hasDocker() {
+		return nil, fmt.Errorf("no Docker client available for running the Cloud Spanner emulator container")
+	}
+	if !isDockerDaemonRunning() {
+		return nil, fmt.Errorf("the Docker daemon does not seem to be running")
+	}
+	const image = "gcr.io/cloud-spanner-emulator/emulator:latest"
+	cmd := sg.Command(ctx, "docker", "pull", image)
+	cmd.Stdout, cmd.Stderr = nil, nil
+	if err := cmd.Run(); err != nil {
+		return nil, err
+	}
+	var dockerRunCmd *exec.Cmd
+	if isRunningOnCloudBuild(ctx) {
+		dockerRunCmd = sg.Command(ctx, "docker", "run", "-d", "--network", "cloudbuild", "--publish-all", image)
+	} else {
+		dockerRunCmd = sg.Command(ctx, "docker", "run", "-d", "--publish-all", image)
+	}
+	var dockerRunStdout strings.Builder
+	dockerRunCmd.Stdout = &dockerRunStdout
+	if err := dockerRunCmd.Run(); err != nil {
+		return nil, err
+	}
+	containerID := strings.TrimSpace(dockerRunStdout.String())
+	cleanup := func() {
```

Comment:
Could we defer cleanups and do an if check on the err named return?

Reply:
I think the readability benefit would be marginal - would prefer to keep as is

Reply:
Alright! Still a little green on go conventions!